From 7990034c06db7a525a7d45a1095ac7ea9cf90a12 Mon Sep 17 00:00:00 2001
From: Uri Mashiach <uri.mashiach@compulab.co.il>
Date: Tue, 19 Feb 2019 21:46:54 +0200
Subject: [PATCH 12/25] mx8m-mini: set Ethernet MAC address environment

If the environment variable "ethaddr" is empty:
* Read the MAC address from the EEPROM.
* Store the MAC address in the environment variable "ethaddr".

Signed-off-by: Uri Mashiach <uri.mashiach@compulab.co.il>
---
 board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c | 37 ++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
index ef5d1ff..80d9a5b 100644
--- a/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
+++ b/board/compulab/ucm-imx8m-mini/ucm-imx8m-mini.c
@@ -24,6 +24,7 @@
 #include <power/pmic.h>
 #include <power/bd71837.h>
 #include "../common/tcpc.h"
+#include "../common/eeprom.h"
 #include <usb.h>
 #include <sec_mipi_dsim.h>
 #include <imx_mipi_dsi_bridge.h>
@@ -183,6 +184,35 @@ static void setup_iomux_fec(void)
 	gpio_direction_output(FEC_RST_PAD, 1);
 }
 
+/*
+ * setup_mac_address() - set Ethernet MAC address environment.
+ *
+ * @return: 0 on success, -1 on failure
+ */
+static int setup_mac_address(void)
+{
+        int ret;
+        unsigned char enetaddr[6];
+
+        ret = eth_env_get_enetaddr("ethaddr", enetaddr);
+        if (ret)
+                return 0;
+
+        ret = cl_eeprom_read_mac_addr(enetaddr, CONFIG_SYS_I2C_EEPROM_BUS);
+        if (ret)
+                return ret;
+
+        ret = is_valid_ethaddr(enetaddr);
+        if (!ret)
+                return -1;
+
+	ret = eth_env_set_enetaddr("ethaddr", enetaddr);
+	if (!ret)
+		return -1;
+
+        return 0;
+}
+
 static int setup_fec(void)
 {
 	struct iomuxc_gpr_base_regs *const iomuxc_gpr_regs
@@ -389,6 +419,7 @@ int board_ehci_usb_phy_mode(struct udevice *dev)
 
 int board_init(void)
 {
+
 #ifdef CONFIG_USB_TCPC
 	setup_typec();
 #endif
@@ -717,10 +748,16 @@ size_t display_count = ARRAY_SIZE(displays);
 
 int board_late_init(void)
 {
+	int ret;
+
 #ifdef CONFIG_ENV_IS_IN_MMC
 	board_late_mmc_env_init();
 #endif
 
+	ret = setup_mac_address();
+	if (ret < 0)
+		printf("%s: Can't set MAC address\n", __func__);
+
 	return 0;
 }
 
-- 
1.9.1

